### PR TITLE
Add template types to Result

### PR DESCRIPTION
**Pull request by paulbalandan**
**Description**
Make ResultInterface a generic type.

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: system/Database/MySQLi/PreparedQuery.php [hunk 1]
```diff
@@ -34,7 +35,7 @@ class PreparedQuery extends BasePreparedQuery
      * @param array $options Passed to the connection's prepare statement.
      *                       Unused in the MySQLi driver.
      *
-     * @return mixed
+     * @return static
```
Comment (kenjis) [inline system/Database/MySQLi/PreparedQuery.php:38]:
```suggestion
     * @return $this
```

File: system/Database/OCI8/PreparedQuery.php [hunk 2]
```diff
@@ -45,7 +45,7 @@ class PreparedQuery extends BasePreparedQuery implements PreparedQueryInterface
      * @param array $options Passed to the connection's prepare statement.
      *                       Unused in the OCI8 driver.
      *
-     * @return mixed
+     * @return static
```
Comment (kenjis) [inline system/Database/OCI8/PreparedQuery.php:48]:
```suggestion
     * @return $this
```

File: system/Database/Postgre/PreparedQuery.php [hunk 3]
```diff
@@ -51,7 +51,7 @@ class PreparedQuery extends BasePreparedQuery
      * @param array $options Passed to the connection's prepare statement.
      *                       Unused in the MySQLi driver.
      *
-     * @return mixed
+     * @return static
```
Comment (kenjis) [inline system/Database/Postgre/PreparedQuery.php:54]:
```suggestion
     * @return $this
```

File: system/Database/SQLSRV/PreparedQuery.php [hunk 4]
```diff
@@ -51,7 +51,7 @@ public function __construct(Connection $db)
      *
      * @param array $options Options takes an associative array;
      *
-     * @return mixed
+     * @return static
```
Comment (kenjis) [inline system/Database/SQLSRV/PreparedQuery.php:54]:
```suggestion
     * @return $this
```

File: system/Database/SQLite3/PreparedQuery.php [hunk 5]
```diff
@@ -41,7 +42,7 @@ class PreparedQuery extends BasePreparedQuery
      * @param array $options Passed to the connection's prepare statement.
      *                       Unused in the MySQLi driver.
      *
-     * @return $this
+     * @return static
```
Comment (kenjis) [inline system/Database/SQLite3/PreparedQuery.php:45]:
```suggestion
     * @return $this
```

**Comment by kenjis** [comment]
I think if the code is `return $this;`, the phpdoc should be `@return $this`.

> $this: the element to which this type applies is the same exact instance as the current class in the given context. As such, this type is a stricter version of static, because the returned instance must not only be of the same class but also the same instance.
https://github.com/php-fig/fig-standards/blob/master/proposed/phpdoc.md

But if we can set return type, it should be `static` (or `self`).

File: system/Database/MySQLi/PreparedQuery.php (hunk 1 above)
Comment (paulbalandan) [reply system/Database/MySQLi/PreparedQuery.php:38]: I will revert this, following changes in #6750

**Comment by MGatner** [comment]
This one can be added at any point as a bugfix/tech-debt change so no need to rush it for 4.3.

**Review by MGatner — APPROVED** [review]
Looks good! Lots more baseline cleanup, makes me happy 🤗

**Review by kenjis — APPROVED** [review]

**Pull request merged**